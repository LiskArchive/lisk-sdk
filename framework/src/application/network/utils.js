/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

'use strict';

const net = require('net');
const dns = require('dns');

const lookupPromise = (hostname, options) =>
	new Promise((resolve, reject) => {
		dns.lookup(hostname, options, (err, address) => {
			if (err) {
				return reject(err);
			}

			return resolve(address);
		});
	});

const lookupPeersIPs = async (peersList, enabled) => {
	// If peers layer is not enabled there is no need to create the peer's list
	if (!enabled) {
		return [];
	}

	// In case domain names are used, resolve those to IP addresses.
	return Promise.all(
		peersList.map(async peer => {
			if (net.isIPv4(peer.ip)) {
				return peer;
			}

			try {
				const address = await lookupPromise(peer.ip, { family: 4 });
				return {
					...peer,
					ip: address,
				};
			} catch (err) {
				console.error(
					`Failed to resolve peer domain name ${peer.ip} to an IP address`,
				);
				return peer;
			}
		}),
	);
};

module.exports = {
	lookupPeersIPs,
};
