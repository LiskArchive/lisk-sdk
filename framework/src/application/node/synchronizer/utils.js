/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

'use strict';

const { maxBy } = require('lodash');
const { ForkStatus } = require('@liskhq/lisk-bft');

const restoreBlocks = async (blocksModule, processorModule, tx = null) => {
	const tempBlocks = await blocksModule.dataAccess.getTempBlocks(
		{},
		{ sort: 'height:asc', limit: null },
		tx,
	);

	if (!tempBlocks || tempBlocks.length === 0) {
		return false;
	}

	for (const tempBlockEntry of tempBlocks) {
		const tempBlockInstance = await processorModule.deserialize(
			tempBlockEntry.fullBlock,
		);
		await processorModule.processValidated(tempBlockInstance, {
			removeFromTempTable: true,
		});
	}

	return true;
};

const clearBlocksTempTable = storageModule =>
	storageModule.entities.TempBlock.truncate();

const deleteBlocksAfterHeight = async (
	processorModule,
	blocksModule,
	logger,
	desiredHeight,
	backup = false,
) => {
	let { height: currentHeight } = blocksModule.lastBlock;
	logger.debug(
		{ desiredHeight, lastBlockHeight: currentHeight },
		'Deleting blocks after height',
	);
	while (desiredHeight < currentHeight) {
		logger.trace(
			{
				height: blocksModule.lastBlock.height,
				blockId: blocksModule.lastBlock.id,
			},
			'Deleting block and backing it up to temporary table',
		);
		const lastBlock = await processorModule.deleteLastBlock({
			saveTempBlock: backup,
		});
		currentHeight = lastBlock.height;
	}
};

/**
 * Allows to restore blocks if there are blocks left upon startup in temp_blocks table (e.g. node crashed)
 * Depends upon fork choice rule if blocks will be applied
 *
 * 1. Gets all blocks from temp_blocks table
 * 2. Sort blocks according to height as we want to lowest height (the next block to be applied)
 * 3. Uses next block with fork choice rule - if fork status indicates we should switch to different chain
 * we continue applying blocks using the `restoreBlocks` function.
 * Otherwise we truncate the temp_blocks table.
 */
const restoreBlocksUponStartup = async (
	logger,
	blocksModule,
	processorModule,
	storageModule,
) => {
	// Get all blocks and find lowest height (next one to be applied)
	const tempBlocks = await storageModule.entities.TempBlock.get(
		{},
		{
			sort: 'height:asc',
			limit: null,
		},
	);
	const blockLowestHeight = tempBlocks[0];
	const blockHighestHeight = tempBlocks[tempBlocks.length - 1];

	const nextTempBlock = await processorModule.deserialize(
		blockHighestHeight.fullBlock,
	);
	const forkStatus = await processorModule.forkStatus(nextTempBlock);
	const blockHasPriority =
		forkStatus === ForkStatus.DIFFERENT_CHAIN ||
		forkStatus === ForkStatus.VALID_BLOCK;

	// Block in the temp table has preference over current tip of the chain
	if (blockHasPriority) {
		logger.info('Restoring blocks from temporary table');
		await deleteBlocksAfterHeight(
			processorModule,
			blocksModule,
			logger,
			blockLowestHeight.height - 1,
			false,
		);
		// In case fork status is DIFFERENT_CHAIN - try to apply blocks from temp_blocks table
		await restoreBlocks(blocksModule, processorModule);
		logger.info('Chain successfully restored');
	} else {
		// Not different chain - Delete remaining blocks from temp_blocks table
		await clearBlocksTempTable(storageModule);
	}
};

const computeBlockHeightsList = (
	finalizedHeight,
	activeDelegates,
	listSizeLimit,
	currentRound,
) => {
	const startingHeight = Math.max(1, (currentRound - 1) * activeDelegates);
	const heightList = new Array(listSizeLimit)
		.fill(0)
		.map((_, i) => startingHeight - i * activeDelegates)
		.filter(height => height > 0);
	const heightListAfterFinalized = heightList.filter(
		height => height > finalizedHeight,
	);
	return heightList.length !== heightListAfterFinalized.length
		? [...heightListAfterFinalized, finalizedHeight]
		: heightListAfterFinalized;
};

// eslint-disable-next-line class-methods-use-this
const computeLargestSubsetMaxBy = (arrayOfObjects, propertySelectorFunc) => {
	const maximumBy = maxBy(arrayOfObjects, propertySelectorFunc);
	const absoluteMax = propertySelectorFunc(maximumBy);
	const largestSubset = [];
	// eslint-disable-next-line no-restricted-syntax
	for (const item of arrayOfObjects) {
		if (propertySelectorFunc(item) === absoluteMax) {
			largestSubset.push(item);
		}
	}
	return largestSubset;
};

module.exports = {
	computeBlockHeightsList,
	computeLargestSubsetMaxBy,
	deleteBlocksAfterHeight,
	restoreBlocksUponStartup,
	restoreBlocks,
	clearBlocksTempTable,
};
