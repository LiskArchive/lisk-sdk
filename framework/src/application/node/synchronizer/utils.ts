/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { ForkStatus, BFT } from '@liskhq/lisk-bft';
import { Chain } from '@liskhq/lisk-chain';
import { Processor } from '../processor';
import { Logger } from '../../logger';

export const restoreBlocks = async (
	chainModule: Chain,
	processorModule: Processor,
): Promise<boolean> => {
	const tempBlocks = await chainModule.dataAccess.getTempBlocks();

	if (tempBlocks.length === 0) {
		return false;
	}

	for (const tempBlock of tempBlocks.reverse()) {
		await processorModule.processValidated(tempBlock, {
			removeFromTempTable: true,
		});
	}

	return true;
};

export const clearBlocksTempTable = async (chainModule: Chain): Promise<void> =>
	chainModule.dataAccess.clearTempBlocks();

export const deleteBlocksAfterHeight = async (
	processorModule: Processor,
	chainModule: Chain,
	logger: Logger,
	desiredHeight: number,
	backup = false,
): Promise<void> => {
	let { height: currentHeight } = chainModule.lastBlock.header;
	logger.debug({ desiredHeight, lastBlockHeight: currentHeight }, 'Deleting blocks after height');
	while (desiredHeight < currentHeight) {
		logger.trace(
			{
				height: chainModule.lastBlock.header.height,
				blockId: chainModule.lastBlock.header.id,
			},
			'Deleting block and backing it up to temporary table',
		);
		await processorModule.deleteLastBlock({
			saveTempBlock: backup,
		});
		currentHeight = chainModule.lastBlock.header.height;
	}
};

/**
 * Allows to restore blocks if there are blocks left upon startup in temp_blocks table (e.g. node crashed)
 * Depends upon fork choice rule if blocks will be applied
 *
 * 1. Gets all blocks from temp_blocks table
 * 2. Sort blocks according to height as we want to lowest height (the next block to be applied)
 * 3. Uses next block with fork choice rule - if fork status indicates we should switch to different chain
 * we continue applying blocks using the `restoreBlocks` function.
 * Otherwise we truncate the temp_blocks table.
 */
export const restoreBlocksUponStartup = async (
	logger: Logger,
	chainModule: Chain,
	bftModule: BFT,
	processorModule: Processor,
): Promise<void> => {
	// Get all blocks and find lowest height (next one to be applied), as it should return in height desc
	const tempBlocks = await chainModule.dataAccess.getTempBlocks();
	const blockLowestHeight = tempBlocks[tempBlocks.length - 1];
	const blockHighestHeight = tempBlocks[0];

	const forkStatus = bftModule.forkChoice(blockHighestHeight.header, chainModule.lastBlock.header);
	const blockHasPriority =
		forkStatus === ForkStatus.DIFFERENT_CHAIN || forkStatus === ForkStatus.VALID_BLOCK;

	// Block in the temp table has preference over current tip of the chain
	if (blockHasPriority) {
		logger.info('Restoring blocks from temporary table');
		await deleteBlocksAfterHeight(
			processorModule,
			chainModule,
			logger,
			blockLowestHeight.header.height - 1,
			false,
		);
		// In case fork status is DIFFERENT_CHAIN - try to apply blocks from temp_blocks table
		await restoreBlocks(chainModule, processorModule);
		logger.info('Chain successfully restored');
	} else {
		// Not different chain - Delete remaining blocks from temp_blocks table
		await clearBlocksTempTable(chainModule);
	}
};

export const computeBlockHeightsList = (
	finalizedHeight: number,
	activeDelegates: number,
	listSizeLimit: number,
	currentRound: number,
): number[] => {
	const startingHeight = Math.max((currentRound - 1) * activeDelegates, 0);
	const heightList = new Array(listSizeLimit)
		.fill(0)
		.map((_, i) => startingHeight - i * activeDelegates)
		.filter(height => height >= 0);
	const heightListAfterFinalized = heightList.filter(height => height > finalizedHeight);
	return heightList.length !== heightListAfterFinalized.length
		? [...heightListAfterFinalized, finalizedHeight]
		: heightListAfterFinalized;
};

export const computeLargestSubsetMaxBy = <T>(
	arrayOfObjects: T[],
	propertySelectorFunc: (param: T) => number,
): T[] => {
	const comparableValues = arrayOfObjects.map(propertySelectorFunc);
	const absoluteMax = Math.max(...comparableValues);
	const largestSubset = [];
	// eslint-disable-next-line no-restricted-syntax
	for (const item of arrayOfObjects) {
		if (propertySelectorFunc(item) === absoluteMax) {
			largestSubset.push(item);
		}
	}
	return largestSubset;
};
