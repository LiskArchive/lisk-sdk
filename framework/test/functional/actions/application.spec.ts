/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { createApplication, closeApplication } from '../utils/application';

import { Application } from '../../../src';

describe('Application related actions', () => {
	let app: Application;

	beforeAll(async () => {
		app = await createApplication('actions-transactions');
	});

	afterAll(async () => {
		await closeApplication(app);
	});

	describe('getSchema', () => {
		it('should return schemas used to encode objects in framework', async () => {
			const frameworkSchemas = await app['_channel'].invoke('app:getSchema');
			expect(frameworkSchemas).toMatchSnapshot();
		});
	});

	describe('getNodeInfo', () => {
		it('should return node status and constants', async () => {
			const nodeStatusAndConstants = await app['_channel'].invoke('app:getNodeInfo');
			expect(nodeStatusAndConstants).toEqual(
				expect.objectContaining({
					height: expect.any(Number),
					version: expect.any(String),
					chainID: expect.any(String),
					networkVersion: expect.any(String),
					lastBlockID: expect.any(String),
					finalizedHeight: expect.any(Number),
					unconfirmedTransactions: expect.any(Number),
				}),
			);
		});
	});

	describe('getForgers', () => {
		it('should return a list of forgers', async () => {
			const forgersInfo = await app['_channel'].invoke<{ [key: string]: any }[]>('app:getForgers');
			expect(forgersInfo).toBeArray();
			forgersInfo.map(forgerInfo => {
				expect(forgerInfo.address).toBeString();
				expect(forgerInfo.nextForgingTime).toBeNumber();
				expect(forgerInfo.minActiveHeight).toBeNumber();
				return expect(forgerInfo.isConsensusParticipant).toBeBoolean();
			});
		});
	});
});
