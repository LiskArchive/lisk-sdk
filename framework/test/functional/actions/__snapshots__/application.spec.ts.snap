// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application related actions getSchema should return schemas used to encode objects in framework 1`] = `
Object {
  "account": Object {
    "$id": "/account/base",
    "properties": Object {
      "address": Object {
        "dataType": "bytes",
        "fieldNumber": 1,
      },
      "pos": Object {
        "fieldNumber": 5,
        "properties": Object {
          "validator": Object {
            "fieldNumber": 1,
            "properties": Object {
              "consecutiveMissedBlocks": Object {
                "dataType": "uint32",
                "fieldNumber": 3,
              },
              "isBanned": Object {
                "dataType": "boolean",
                "fieldNumber": 5,
              },
              "lastForgedHeight": Object {
                "dataType": "uint32",
                "fieldNumber": 4,
              },
              "pomHeights": Object {
                "fieldNumber": 2,
                "items": Object {
                  "dataType": "uint32",
                },
                "type": "array",
              },
              "totalStakeReceived": Object {
                "dataType": "uint64",
                "fieldNumber": 6,
              },
              "username": Object {
                "dataType": "string",
                "fieldNumber": 1,
              },
            },
            "required": Array [
              "username",
              "pomHeights",
              "consecutiveMissedBlocks",
              "lastForgedHeight",
              "isBanned",
              "totalStakeReceived",
            ],
            "type": "object",
          },
          "sentStakes": Object {
            "fieldNumber": 2,
            "items": Object {
              "properties": Object {
                "amount": Object {
                  "dataType": "uint64",
                  "fieldNumber": 2,
                },
                "validatorAddress": Object {
                  "dataType": "bytes",
                  "fieldNumber": 1,
                },
              },
              "required": Array [
                "validatorAddress",
                "amount",
              ],
              "type": "object",
            },
            "type": "array",
          },
          "unlocking": Object {
            "fieldNumber": 3,
            "items": Object {
              "properties": Object {
                "amount": Object {
                  "dataType": "uint64",
                  "fieldNumber": 2,
                },
                "validatorAddress": Object {
                  "dataType": "bytes",
                  "fieldNumber": 1,
                },
                "unstakeHeight": Object {
                  "dataType": "uint32",
                  "fieldNumber": 3,
                },
              },
              "required": Array [
                "validatorAddress",
                "amount",
                "unstakeHeight",
              ],
              "type": "object",
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "keys": Object {
        "fieldNumber": 4,
        "properties": Object {
          "mandatoryKeys": Object {
            "fieldNumber": 2,
            "items": Object {
              "dataType": "bytes",
            },
            "type": "array",
          },
          "numberOfSignatures": Object {
            "dataType": "uint32",
            "fieldNumber": 1,
          },
          "optionalKeys": Object {
            "fieldNumber": 3,
            "items": Object {
              "dataType": "bytes",
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "sequence": Object {
        "fieldNumber": 3,
        "properties": Object {
          "nonce": Object {
            "dataType": "uint64",
            "fieldNumber": 1,
          },
        },
        "type": "object",
      },
      "token": Object {
        "fieldNumber": 2,
        "properties": Object {
          "balance": Object {
            "dataType": "uint64",
            "fieldNumber": 1,
          },
        },
        "type": "object",
      },
    },
    "required": Array [
      "address",
      "token",
      "sequence",
      "keys",
      "pos",
    ],
    "type": "object",
  },
  "block": Object {
    "$id": "/block",
    "properties": Object {
      "header": Object {
        "dataType": "bytes",
        "fieldNumber": 1,
      },
      "transactions": Object {
        "fieldNumber": 2,
        "items": Object {
          "dataType": "bytes",
        },
        "type": "array",
      },
    },
    "required": Array [
      "header",
      "transactions",
    ],
    "type": "object",
  },
  "blockHeader": Object {
    "$id": "/block/header",
    "properties": Object {
      "asset": Object {
        "dataType": "bytes",
        "fieldNumber": 8,
      },
      "generatorPublicKey": Object {
        "dataType": "bytes",
        "fieldNumber": 6,
      },
      "height": Object {
        "dataType": "uint32",
        "fieldNumber": 3,
      },
      "previousBlockID": Object {
        "dataType": "bytes",
        "fieldNumber": 4,
      },
      "reward": Object {
        "dataType": "uint64",
        "fieldNumber": 7,
      },
      "signature": Object {
        "dataType": "bytes",
        "fieldNumber": 9,
      },
      "timestamp": Object {
        "dataType": "uint32",
        "fieldNumber": 2,
      },
      "transactionRoot": Object {
        "dataType": "bytes",
        "fieldNumber": 5,
      },
      "version": Object {
        "dataType": "uint32",
        "fieldNumber": 1,
      },
    },
    "required": Array [
      "version",
      "timestamp",
      "height",
      "previousBlockID",
      "transactionRoot",
      "generatorPublicKey",
      "reward",
      "asset",
    ],
    "type": "object",
  },
  "blockHeadersAssets": Object {
    "0": Object {
      "$id": "/genesisBlock/header/asset",
      "properties": Object {
        "accounts": Object {
          "fieldNumber": 1,
          "items": Object {
            "$id": "/account/base",
            "properties": Object {
              "address": Object {
                "dataType": "bytes",
                "fieldNumber": 1,
              },
              "pos": Object {
                "fieldNumber": 5,
                "properties": Object {
                  "validator": Object {
                    "fieldNumber": 1,
                    "properties": Object {
                      "consecutiveMissedBlocks": Object {
                        "dataType": "uint32",
                        "fieldNumber": 3,
                      },
                      "isBanned": Object {
                        "dataType": "boolean",
                        "fieldNumber": 5,
                      },
                      "lastForgedHeight": Object {
                        "dataType": "uint32",
                        "fieldNumber": 4,
                      },
                      "pomHeights": Object {
                        "fieldNumber": 2,
                        "items": Object {
                          "dataType": "uint32",
                        },
                        "type": "array",
                      },
                      "totalStakeReceived": Object {
                        "dataType": "uint64",
                        "fieldNumber": 6,
                      },
                      "username": Object {
                        "dataType": "string",
                        "fieldNumber": 1,
                      },
                    },
                    "required": Array [
                      "username",
                      "pomHeights",
                      "consecutiveMissedBlocks",
                      "lastForgedHeight",
                      "isBanned",
                      "totalStakeReceived",
                    ],
                    "type": "object",
                  },
                  "sentStakes": Object {
                    "fieldNumber": 2,
                    "items": Object {
                      "properties": Object {
                        "amount": Object {
                          "dataType": "uint64",
                          "fieldNumber": 2,
                        },
                        "validatorAddress": Object {
                          "dataType": "bytes",
                          "fieldNumber": 1,
                        },
                      },
                      "required": Array [
                        "validatorAddress",
                        "amount",
                      ],
                      "type": "object",
                    },
                    "type": "array",
                  },
                  "unlocking": Object {
                    "fieldNumber": 3,
                    "items": Object {
                      "properties": Object {
                        "amount": Object {
                          "dataType": "uint64",
                          "fieldNumber": 2,
                        },
                        "validatorAddress": Object {
                          "dataType": "bytes",
                          "fieldNumber": 1,
                        },
                        "unstakeHeight": Object {
                          "dataType": "uint32",
                          "fieldNumber": 3,
                        },
                      },
                      "required": Array [
                        "validatorAddress",
                        "amount",
                        "unstakeHeight",
                      ],
                      "type": "object",
                    },
                    "type": "array",
                  },
                },
                "type": "object",
              },
              "keys": Object {
                "fieldNumber": 4,
                "properties": Object {
                  "mandatoryKeys": Object {
                    "fieldNumber": 2,
                    "items": Object {
                      "dataType": "bytes",
                    },
                    "type": "array",
                  },
                  "numberOfSignatures": Object {
                    "dataType": "uint32",
                    "fieldNumber": 1,
                  },
                  "optionalKeys": Object {
                    "fieldNumber": 3,
                    "items": Object {
                      "dataType": "bytes",
                    },
                    "type": "array",
                  },
                },
                "type": "object",
              },
              "sequence": Object {
                "fieldNumber": 3,
                "properties": Object {
                  "nonce": Object {
                    "dataType": "uint64",
                    "fieldNumber": 1,
                  },
                },
                "type": "object",
              },
              "token": Object {
                "fieldNumber": 2,
                "properties": Object {
                  "balance": Object {
                    "dataType": "uint64",
                    "fieldNumber": 1,
                  },
                },
                "type": "object",
              },
            },
            "required": Array [
              "address",
              "token",
              "sequence",
              "keys",
              "pos",
            ],
            "type": "object",
          },
          "type": "array",
        },
        "initValidators": Object {
          "fieldNumber": 2,
          "items": Object {
            "dataType": "bytes",
          },
          "minItems": 1,
          "type": "array",
        },
        "initRounds": Object {
          "dataType": "uint32",
          "fieldNumber": 3,
          "minimum": 3,
        },
      },
      "required": Array [
        "accounts",
        "initValidators",
        "initRounds",
      ],
      "type": "object",
    },
    "2": Object {
      "$id": "/blockHeader/asset/v2",
      "properties": Object {
        "maxHeightPreviouslyForged": Object {
          "dataType": "uint32",
          "fieldNumber": 1,
        },
        "maxHeightPrevoted": Object {
          "dataType": "uint32",
          "fieldNumber": 2,
        },
        "seedReveal": Object {
          "dataType": "bytes",
          "fieldNumber": 3,
          "maxLength": 16,
          "minLength": 16,
        },
      },
      "required": Array [
        "maxHeightPreviouslyForged",
        "maxHeightPrevoted",
        "seedReveal",
      ],
      "type": "object",
    },
  },
  "transaction": Object {
    "$id": "lisk/transaction",
    "properties": Object {
      "asset": Object {
        "dataType": "bytes",
        "fieldNumber": 6,
      },
      "assetID": Object {
        "dataType": "uint32",
        "fieldNumber": 2,
      },
      "fee": Object {
        "dataType": "uint64",
        "fieldNumber": 4,
      },
      "moduleID": Object {
        "dataType": "uint32",
        "fieldNumber": 1,
        "minimum": 2,
      },
      "nonce": Object {
        "dataType": "uint64",
        "fieldNumber": 3,
      },
      "senderPublicKey": Object {
        "dataType": "bytes",
        "fieldNumber": 5,
        "maxLength": 32,
        "minLength": 32,
      },
      "signatures": Object {
        "fieldNumber": 7,
        "items": Object {
          "dataType": "bytes",
        },
        "type": "array",
      },
    },
    "required": Array [
      "moduleID",
      "assetID",
      "nonce",
      "fee",
      "senderPublicKey",
      "asset",
    ],
    "type": "object",
  },
  "transactionsAssets": Array [
    Object {
      "assetID": 0,
      "assetName": "transfer",
      "moduleID": 2,
      "moduleName": "token",
      "schema": Object {
        "$id": "lisk/transfer-asset",
        "properties": Object {
          "amount": Object {
            "dataType": "uint64",
            "fieldNumber": 1,
          },
          "data": Object {
            "dataType": "string",
            "fieldNumber": 3,
          },
          "recipientAddress": Object {
            "dataType": "bytes",
            "fieldNumber": 2,
          },
        },
        "required": Array [
          "amount",
          "recipientAddress",
          "data",
        ],
        "title": "Transfer transaction asset",
        "type": "object",
      },
    },
    Object {
      "assetID": 0,
      "assetName": "registerMultisignature",
      "moduleID": 4,
      "moduleName": "keys",
      "schema": Object {
        "$id": "lisk/keys/register",
        "properties": Object {
          "mandatoryKeys": Object {
            "fieldNumber": 2,
            "items": Object {
              "dataType": "bytes",
              "maxLength": 32,
              "minLength": 32,
            },
            "maxItems": 64,
            "minItems": 0,
            "type": "array",
          },
          "numberOfSignatures": Object {
            "dataType": "uint32",
            "fieldNumber": 1,
            "maximum": 64,
            "minimum": 1,
          },
          "optionalKeys": Object {
            "fieldNumber": 3,
            "items": Object {
              "dataType": "bytes",
              "maxLength": 32,
              "minLength": 32,
            },
            "maxItems": 64,
            "minItems": 0,
            "type": "array",
          },
        },
        "required": Array [
          "numberOfSignatures",
          "optionalKeys",
          "mandatoryKeys",
        ],
        "type": "object",
      },
    },
    Object {
      "assetID": 0,
      "assetName": "registerValidator",
      "moduleID": 5,
      "moduleName": "pos",
      "schema": Object {
        "$id": "lisk/pos/register",
        "properties": Object {
          "username": Object {
            "dataType": "string",
            "fieldNumber": 1,
            "maxLength": 20,
            "minLength": 1,
          },
        },
        "required": Array [
          "username",
        ],
        "type": "object",
      },
    },
    Object {
      "assetID": 1,
      "assetName": "stake",
      "moduleID": 5,
      "moduleName": "pos",
      "schema": Object {
        "$id": "lisk/pos/stake",
        "properties": Object {
          "stakes": Object {
            "fieldNumber": 1,
            "items": Object {
              "properties": Object {
                "amount": Object {
                  "dataType": "sint64",
                  "fieldNumber": 2,
                },
                "validatorAddress": Object {
                  "dataType": "bytes",
                  "fieldNumber": 1,
                  "maxLength": 20,
                  "minLength": 20,
                },
              },
              "required": Array [
                "validatorAddress",
                "amount",
              ],
              "type": "object",
            },
            "maxItems": 20,
            "minItems": 1,
            "type": "array",
          },
        },
        "required": Array [
          "stakes",
        ],
        "type": "object",
      },
    },
    Object {
      "assetID": 2,
      "assetName": "unlockToken",
      "moduleID": 5,
      "moduleName": "pos",
      "schema": Object {
        "$id": "lisk/pos/unlock",
        "properties": Object {
          "unlockObjects": Object {
            "fieldNumber": 1,
            "items": Object {
              "properties": Object {
                "amount": Object {
                  "dataType": "uint64",
                  "fieldNumber": 2,
                },
                "validatorAddress": Object {
                  "dataType": "bytes",
                  "fieldNumber": 1,
                  "maxLength": 20,
                  "minLength": 20,
                },
                "unstakeHeight": Object {
                  "dataType": "uint32",
                  "fieldNumber": 3,
                },
              },
              "required": Array [
                "validatorAddress",
                "amount",
                "unstakeHeight",
              ],
              "type": "object",
            },
            "maxItems": 20,
            "minItems": 1,
            "type": "array",
          },
        },
        "required": Array [
          "unlockObjects",
        ],
        "type": "object",
      },
    },
    Object {
      "assetID": 3,
      "assetName": "reportDelegateMisbehavior",
      "moduleID": 5,
      "moduleName": "pos",
      "schema": Object {
        "$id": "lisk/pos/pom",
        "properties": Object {
          "header1": Object {
            "$id": "block-header1",
            "fieldNumber": 1,
            "properties": Object {
              "asset": Object {
                "fieldNumber": 8,
                "properties": Object {
                  "maxHeightPreviouslyForged": Object {
                    "dataType": "uint32",
                    "fieldNumber": 1,
                  },
                  "maxHeightPrevoted": Object {
                    "dataType": "uint32",
                    "fieldNumber": 2,
                  },
                  "seedReveal": Object {
                    "dataType": "bytes",
                    "fieldNumber": 3,
                  },
                },
                "required": Array [
                  "maxHeightPreviouslyForged",
                  "maxHeightPrevoted",
                  "seedReveal",
                ],
                "type": "object",
              },
              "generatorPublicKey": Object {
                "dataType": "bytes",
                "fieldNumber": 6,
              },
              "height": Object {
                "dataType": "uint32",
                "fieldNumber": 3,
              },
              "previousBlockID": Object {
                "dataType": "bytes",
                "fieldNumber": 4,
              },
              "reward": Object {
                "dataType": "uint64",
                "fieldNumber": 7,
              },
              "signature": Object {
                "dataType": "bytes",
                "fieldNumber": 9,
              },
              "timestamp": Object {
                "dataType": "uint32",
                "fieldNumber": 2,
              },
              "transactionRoot": Object {
                "dataType": "bytes",
                "fieldNumber": 5,
              },
              "version": Object {
                "dataType": "uint32",
                "fieldNumber": 1,
              },
            },
            "required": Array [
              "version",
              "timestamp",
              "height",
              "previousBlockID",
              "transactionRoot",
              "generatorPublicKey",
              "reward",
              "asset",
            ],
            "type": "object",
          },
          "header2": Object {
            "$id": "block-header2",
            "fieldNumber": 2,
            "properties": Object {
              "asset": Object {
                "fieldNumber": 8,
                "properties": Object {
                  "maxHeightPreviouslyForged": Object {
                    "dataType": "uint32",
                    "fieldNumber": 1,
                  },
                  "maxHeightPrevoted": Object {
                    "dataType": "uint32",
                    "fieldNumber": 2,
                  },
                  "seedReveal": Object {
                    "dataType": "bytes",
                    "fieldNumber": 3,
                  },
                },
                "required": Array [
                  "maxHeightPreviouslyForged",
                  "maxHeightPrevoted",
                  "seedReveal",
                ],
                "type": "object",
              },
              "generatorPublicKey": Object {
                "dataType": "bytes",
                "fieldNumber": 6,
              },
              "height": Object {
                "dataType": "uint32",
                "fieldNumber": 3,
              },
              "previousBlockID": Object {
                "dataType": "bytes",
                "fieldNumber": 4,
              },
              "reward": Object {
                "dataType": "uint64",
                "fieldNumber": 7,
              },
              "signature": Object {
                "dataType": "bytes",
                "fieldNumber": 9,
              },
              "timestamp": Object {
                "dataType": "uint32",
                "fieldNumber": 2,
              },
              "transactionRoot": Object {
                "dataType": "bytes",
                "fieldNumber": 5,
              },
              "version": Object {
                "dataType": "uint32",
                "fieldNumber": 1,
              },
            },
            "required": Array [
              "version",
              "timestamp",
              "height",
              "previousBlockID",
              "transactionRoot",
              "generatorPublicKey",
              "reward",
              "asset",
            ],
            "type": "object",
          },
        },
        "required": Array [
          "header1",
          "header2",
        ],
        "type": "object",
      },
    },
  ],
}
`;
